lib/backend/postgres/azure.go

```diff
@@ -0,0 +1,63 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package postgres
+
+import (
+	"context"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/gravitational/trace"
+	"github.com/jackc/pgx/v4"
+	"github.com/sirupsen/logrus"
+)
+
+func azureBeforeConnect(clientID string, log logrus.FieldLogger) (func(ctx context.Context, config *pgx.ConnConfig) error, error) {
+	var cred azcore.TokenCredential
+	if clientID != "" {
+		c, err := azidentity.NewManagedIdentityCredential(&azidentity.ManagedIdentityCredentialOptions{
+			ID: azidentity.ClientID(clientID),
+		})
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		cred = c
+	} else {
+		c, err := azidentity.NewDefaultAzureCredential(nil)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		cred = c
+	}
+
+	beforeConnect := func(ctx context.Context, config *pgx.ConnConfig) error {
+		log.Debugf("Fetching Azure access token.")
+		token, err := cred.GetToken(ctx, policy.TokenRequestOptions{
+			Scopes: []string{"https://ossrdbms-aad.database.windows.net/.default"},
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		log.WithField("ttl", time.Until(token.ExpiresOn).String()).Debugf("Fetched Azure access token.")
```

Comment:
nit:

```suggestion
		log.WithField("ttl", time.Until(token.ExpiresOn).String()).Debug("Fetched Azure access token.")
```

Where we are doing proper structured logging, it would be nice if we tried to use `Debug` rather than `Debugf` since this will make it easier for us to identify places that can be converted to use structured logging in future :)

Reply:
Added some more logging, all without string formatting.